files/scripts/supervisor-proc-exit-listener

```diff
@@ -62,51 +157,54 @@ def main(argv):
-                sys.exit(3)
-
-            restart_feature = features_table[container_name].get('auto_restart')
-            if not restart_feature:
-                syslog.syslog(
-                    syslog.LOG_ERR, "Unable to determine auto-restart feature status for '{}'. Exiting...".format(container_name))
-                sys.exit(4)
-
-            # If auto-restart feature is not disabled and at the same time
-            # a critical process exited unexpectedly, terminate supervisor
-            if (restart_feature != 'disabled' and expected == 0 and
-                    (processname in critical_process_list or groupname in critical_group_list)):
-                MSG_FORMAT_STR = "Process {} exited unxepectedly. Terminating supervisor..."
-                msg = MSG_FORMAT_STR.format(payload_headers['processname'])
-                syslog.syslog(syslog.LOG_INFO, msg)
-                os.kill(os.getppid(), signal.SIGTERM)
+        if not listener_state_ready:
+            # Transition from ACKNOWLEDGED to READY
+            childutils.listener.ready()
+            listener_state_ready = True
+
+        file_descriptor_list = select.select([sys.stdin], [], [], SELECT_TIMEOUT_SECS)[0]
+        if len(file_descriptor_list) > 0:
+            line = file_descriptor_list[0].readline()
+            headers = childutils.get_headers(line)
+            payload = sys.stdin.read(int(headers['len']))
+
+            # We only care about PROCESS_STATE_EXITED events
+            if headers['eventname'] == 'PROCESS_STATE_EXITED':
+                payload_headers, payload_data = childutils.eventdata(payload + '\n')
+
+                expected = int(payload_headers['expected'])
+                processname = payload_headers['processname']
+                groupname = payload_headers['groupname']
+
+                if ((processname in critical_process_list or groupname in critical_group_list)
+                        and expected == 0):
+                    if container_name != "database":
+                        restart_feature = get_autorestart_state(container_name)
+
+                    if container_name == "database" or restart_feature != "disabled":
+                        MSG_FORMAT_STR = "Process '{}' exited unexpectedly. Terminating supervisor '{}'"
+                        msg = MSG_FORMAT_STR.format(payload_headers['processname'], container_name)
+                        syslog.syslog(syslog.LOG_INFO, msg)
+                        os.kill(os.getppid(), signal.SIGTERM)
+                    else:
+                        process_under_alerting[processname] = time.time()
+
+            # Transition from BUSY to ACKNOWLEDGED
+            childutils.listener.ok()
+            listener_state_ready = False
+
+        # Check whether we need write alerting messages into syslog
+        for process in process_under_alerting.keys():
+            epoch_time = time.time()
+            if process_running(process):
```

Comment:
this is unnecessary, if the process got start, you will receive msg from proc listener, why do you need to use supervisorctl to check?

Reply:
Since this event listener is only interested in the event `PROCESS_STATE_EXITED` (see the line 176), it will not receive the notification from Supervisord about the event `PROCESS_STATE_RUNNING` which was caused by running status of a process changed for example from `EXITED` to `RUNNING`. This is why I used the command `supervisorctl status` to check the status of each process.

Reply:
why not update that? in this case, the proc exit listener should be interested both EXITED, RUNNING state. btw, what are other states?

Reply:
I will update the PR and remove the function `is_process_running(...)`, `get_command_result(...)`.  The event types regarding to the process status are `PROCESS_STATE_RUNNING`, `PROCESS_STATE_STARTING`, `PROCESS_STATE_BACKOFF`, `PROCESS_STATE_EXITED`, `PROCESS_STATE_STOPPING`, `PROCESS_STATE_STOPPED`, `PROCESS_STATE_UNKNOWN` and `PROCESS_STATE_FATAL`.